<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · MendelImpute</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">MendelImpute</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../Phasing_and_Imputation/">Phasing and Imputation</a></li><li><a class="tocitem" href="../performance/">Performance Gotchas</a></li><li><a class="tocitem" href="../painting/">Estimating ancestry</a></li><li><a class="tocitem" href="../ultra+compress/">Ultra compression</a></li><li><a class="tocitem" href="../script/">Run as script</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/MendelImpute.jl/blob/master/docs/src/man/api.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><p>Documentation for <code>MendelImpute.jl</code>&#39;s functions.</p><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#MendelImpute.admixture_global"><code>MendelImpute.admixture_global</code></a></li><li><a href="#MendelImpute.admixture_local"><code>MendelImpute.admixture_local</code></a></li><li><a href="#MendelImpute.compress_haplotypes"><code>MendelImpute.compress_haplotypes</code></a></li><li><a href="#MendelImpute.convert_compressed"><code>MendelImpute.convert_compressed</code></a></li><li><a href="#MendelImpute.phase"><code>MendelImpute.phase</code></a></li><li><a href="#MendelImpute.thousand_genome_population_to_superpopulation"><code>MendelImpute.thousand_genome_population_to_superpopulation</code></a></li><li><a href="#MendelImpute.thousand_genome_samples_to_population"><code>MendelImpute.thousand_genome_samples_to_population</code></a></li><li><a href="#MendelImpute.thousand_genome_samples_to_super_population"><code>MendelImpute.thousand_genome_samples_to_super_population</code></a></li></ul><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.phase" href="#MendelImpute.phase"><code>MendelImpute.phase</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">phase(tgtfile::String, reffile::String, outfile::String; [impute::Bool],
    [phase::Bool], [dosage::Bool], [rescreen::Bool], [max_haplotypes::Int], 
    [stepwise::Int], [thinning_factor::Int], [scale_allelefreq::Bool], 
    [dynamic_programming::Bool])</code></pre><p>Main function of MendelImpute program. Phasing (haplotying) of <code>tgtfile</code> from a pool of haplotypes <code>reffile</code> by sliding windows and saves result in <code>outfile</code>. All SNPs in <code>tgtfile</code> must be present in <code>reffile</code>. Per-SNP quality score will be saved in <code>outfile</code>, while per-sample imputation score will be saved in a file ending in <code>sample.error</code>.</p><p><strong>Input</strong></p><ul><li><code>tgtfile</code>: VCF, PLINK, or BGEN file. VCF files should end in <code>.vcf</code> or <code>.vcf.gz</code>.   PLINK files should exclude <code>.bim/.bed/.fam</code> trailings but the trio must all   be present in the same directory. BGEN files should end in <code>.bgen</code></li><li><code>reffile</code>: Reference haplotype file ending in <code>.jlso</code> (compressed binary files).   See <a href="#MendelImpute.compress_haplotypes"><code>compress_haplotypes</code></a></li><li><code>outfile</code>: output filename ending in <code>.vcf.gz</code>, <code>.vcf</code>, or <code>.jlso</code>. VCF output   genotypes will have no missing data. If ending in <code>.jlso</code>, will output   ultra-compressed data structure recording <code>HaplotypeMosaicPair</code>s for    each sample.</li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>impute</code>: If <code>true</code>, imputes every SNPs in <code>reffile</code> to <code>tgtfile</code>. Otherwise   only missing snps in <code>tgtfile</code> will be imputed.</li><li><code>phase</code>: If <code>true</code>, all output genotypes will be phased, but observed data    (minor allele count) may be changed. If <code>phase=false</code> all output genotypes   will be unphased but observed minor allele count will not change.</li><li><code>dosage</code>: If <code>true</code>, will assume target matrix are dosages for imputation. Note   this means the genotype matrix will be entirely single precision. </li><li><code>rescreen</code>: This option is more computationally intensive but gives more   accurate results. It saves a number of top haplotype pairs when solving   the least squares objective, and re-minimize least squares on just   observed data.</li><li><code>max_haplotypes</code>: Maximum number of haplotypes for using global search. Windows   exceeding this number of unique haplotypes will be searched using a   heuristic. A non-zero <code>stepscreen</code> or <code>thinning_factor</code> need to be specified </li><li><code>stepwise</code>: If an integer is specified, will solve the least squares objective   by first finding <code>stepwise</code> top haplotypes using a stepwise heuristic then   finds the next haplotype using global search. Uses <code>max_haplotypes</code>. </li><li><code>thinning_factor</code>: If an integer is specified, will solve the least squares   objective on only <code>thining_factor</code> unique haplotypes. Uses <code>max_haplotypes</code>.</li><li><code>scale_allelefreq</code>: Boolean indicating whether to give rare SNPs more weight   scaled by <code>wᵢ = 1 / √2p(1-p)</code> where max weight is 2. </li><li><code>dynamic_programming</code>: Boolean indicating whether to phase with a global    search that finds the longest haplotype stretch over all windows. (Currently   broken, sorry!)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/phase.jl#L5-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.compress_haplotypes" href="#MendelImpute.compress_haplotypes"><code>MendelImpute.compress_haplotypes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">compress_haplotypes(reffile::String, tgtfile::String, outfile::String, 
    [d::Int], [minwidth::Int], [overlap::Float64])</code></pre><p>Cuts a haplotype matrix <code>reffile</code> into windows of variable width so that each window has less than <code>d</code> unique haplotypes. Saves result to <code>outfile</code> as a compressed binary format. All SNPs in <code>tgtfile</code> must be present in <code>reffile</code>.  All genotypes in <code>reffile</code> must be phased and non-missing, and all genotype positions must be unique. </p><p><strong>Inputs</strong></p><ul><li><code>reffile</code>: reference haplotype file name (ends in <code>.vcf</code>, <code>.vcf.gz</code>, or <code>.bgen</code>)</li><li><code>tgtfile</code>: VCF, PLINK, or BGEN file. VCF files should end in <code>.vcf</code> or <code>.vcf.gz</code>.   PLINK files should exclude <code>.bim/.bed/.fam</code> suffixes but the trio must all   be present in the same directory. BGEN files should end in <code>.bgen</code>.</li><li><code>outfile</code>: Output file name (ends in <code>.jlso</code>)</li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>d</code>: Max number of unique haplotypes per genotype window (default <code>d = 1000</code>). </li><li><code>minwidth</code>: Minimum number of typed SNPs per window (default 0)</li><li><code>overlap</code>: How much overlap between adjacent genotype windows in percentage of   each window&#39;s width (default 0.0)</li></ul><p><strong>Why is <code>tgtfile</code> required?</strong></p><p>The unique haplotypes in each window is computed on the typed SNPs only.  A genotype matrix <code>tgtfile</code> is used to identify the typed SNPs. In the future,  hopefully we can pre-compute compressed haplotype panels for all genotyping  platforms and provide them as downloadable files. But currently, users must run this function by themselves. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/compress.jl#L278-L307">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.convert_compressed" href="#MendelImpute.convert_compressed"><code>MendelImpute.convert_compressed</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">convert_compressed(t&lt;:Real, phaseinfo::String, reffile::String)</code></pre><p>Converts <code>phaseinfo</code> into a phased genotype matrix of type <code>t</code> using the full reference haplotype panel <code>H</code> </p><p><strong>Inputs</strong></p><ul><li><code>t</code>: Type of matrix. If <code>bool</code>, genotypes are converted to a <code>BitMatrix</code></li><li><code>phaseinfo</code>: Vector of <code>HaplotypeMosaicPair</code>s stored in <code>.jlso</code> format</li><li><code>reffile</code>: The complete (uncompressed) haplotype reference file</li></ul><p><strong>Output</strong></p><ul><li><code>X1</code>: allele 1 of the phased genotype. Each column is a sample. <code>X = X1 + X2</code>. </li><li><code>X2</code>: allele 2 of the phased genotype. Each column is a sample. <code>X = X1 + X2</code>. </li><li><code>phase</code>: the original data structure after phasing and imputation.</li><li><code>sampleID</code>: The ID&#39;s of each imputed person.  </li><li><code>H</code>: the complete reference haplotype panel. Columns of <code>H</code> are haplotypes.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/ultra_compress.jl#L5-L22">source</a></section><section><div><pre><code class="nohighlight hljs">convert_compressed(t&lt;:Real, phaseinfo::Vector{HaplotypeMosaicPair}, H::AbstractMatrix)</code></pre><p>Columns of <code>H</code> are haplotypes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/ultra_compress.jl#L39-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.admixture_global" href="#MendelImpute.admixture_global"><code>MendelImpute.admixture_global</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">admixture_global(tgtfile::String, reffile::String, 
    refID_to_population::Dict{String, String}, populations::Vector{String})</code></pre><p>Computes global ancestry estimates for each sample in <code>tgtfile</code> using a labeled reference panel <code>reffile</code>. </p><p><strong>Inputs</strong></p><ul><li><code>tgtfile</code>: VCF or PLINK files. VCF files should end in <code>.vcf</code> or <code>.vcf.gz</code>.   PLINK files should exclude <code>.bim/.bed/.fam</code> trailings but the trio must all   be present in the same directory.</li><li><code>reffile</code>: Reference haplotype file ending in <code>.jlso</code> (compressed binary files).   See <a href="#MendelImpute.compress_haplotypes"><code>compress_haplotypes</code></a>.</li><li><code>refID_to_population</code>: A dictionary mapping each sample IDs in the haplotype    reference panel to their population origin. For examples, see output of   <a href="#MendelImpute.thousand_genome_population_to_superpopulation"><code>thousand_genome_population_to_superpopulation</code></a> and   <a href="#MendelImpute.thousand_genome_samples_to_super_population"><code>thousand_genome_samples_to_super_population</code></a></li><li><code>populations</code>: A vector of <code>String</code> containing unique populations present in   <code>values(refID_to_population)</code>. </li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>Q_outfile</code>: Output file name for the estimated <code>Q</code> matrix. Default   <code>Q_outfile=&quot;mendelimpute.ancestry.Q&quot;</code>.</li><li><code>imputed_outfile</code>: Output file name for the imputed genotypes ending in <code>.jlso</code>.   Default <code>impute_outfile = &quot;mendelimpute.ancestry.Q.jlso&quot;</code></li></ul><p><strong>Output</strong></p><ul><li><code>Q</code>: A <code>DataFrame</code> containing estimated ancestry fractions. Each row is a sample.   Matrix will be saved in <code>mendelimpute.ancestry.Q</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/ancestry.jl#L5-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.admixture_local" href="#MendelImpute.admixture_local"><code>MendelImpute.admixture_local</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">admixture_local(tgtfile::String, reffile::String, 
    refID_to_population::Dict{String, String}, populations::Vector{String},
    population_colors::Vector{RGB{FixedPointNumbers.N0f8}})</code></pre><p>Computes global ancestry estimates for each sample in <code>tgtfile</code> using a labeled reference panel <code>reffile</code>. </p><p><strong>Inputs</strong></p><ul><li><code>tgtfile</code>: VCF or PLINK files. VCF files should end in <code>.vcf</code> or <code>.vcf.gz</code>.   PLINK files should exclude <code>.bim/.bed/.fam</code> trailings but the trio must all   be present in the same directory.</li><li><code>reffile</code>: Reference haplotype file ending in <code>.jlso</code> (compressed binary files).   See <a href="#MendelImpute.compress_haplotypes"><code>compress_haplotypes</code></a>.</li><li><code>refID_to_population</code>: A dictionary mapping each sample IDs in the haplotype    reference panel to their population origin. For examples, see output of   <a href="#MendelImpute.thousand_genome_population_to_superpopulation"><code>thousand_genome_population_to_superpopulation</code></a> and   <a href="#MendelImpute.thousand_genome_samples_to_super_population"><code>thousand_genome_samples_to_super_population</code></a></li><li><code>population</code>: A list <code>String</code> containing unique populations present in   <code>values(refID_to_population)</code>. </li><li><code>population_colors</code>: A vector of colors for each population.   <code>typeof(population_colors}</code> should be <code>Vector{RGB{FixedPointNumbers.N0f8}}</code></li></ul><p><strong>Output</strong></p><ul><li><code>Q</code>: Matrix containing estimated ancestry fractions. Each row is a haplotype.   Sample 1&#39;s haplotypes are in rows 1 and 2, sample 2&#39;s are in rows 3, 4...etc.</li><li><code>pop_colors</code>: Matrix with sample dimension of <code>Q</code> storing colors. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/ancestry.jl#L65-L92">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.thousand_genome_samples_to_population" href="#MendelImpute.thousand_genome_samples_to_population"><code>MendelImpute.thousand_genome_samples_to_population</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">thousand_genome_samples_to_population()</code></pre><p>Creates a dictionaries mapping sample IDs of 1000 genome project to 26 population codes.</p><p>Population code and super population codes are described here: https://www.internationalgenome.org/category/population/</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/ancestry.jl#L331-L339">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.thousand_genome_samples_to_super_population" href="#MendelImpute.thousand_genome_samples_to_super_population"><code>MendelImpute.thousand_genome_samples_to_super_population</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">thousand_genome_samples_to_population()</code></pre><p>Creates a dictionaries mapping sample IDs of 1000 genome project to 5 super population codes.</p><p>Population code and super population codes are described here: https://www.internationalgenome.org/category/population/</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/ancestry.jl#L353-L361">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelImpute.thousand_genome_population_to_superpopulation" href="#MendelImpute.thousand_genome_population_to_superpopulation"><code>MendelImpute.thousand_genome_population_to_superpopulation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">thousand_genome_population_to_superpopulation()</code></pre><p>Creates a dictionary mapping population codes of 1000 genome project to their super-population codes.</p><p>Population code and super population codes are described here: https://www.internationalgenome.org/category/population/</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelImpute.jl/blob/860ce2fc59a30ec536bc154d9d048268c2cba25e/src/ancestry.jl#L378-L386">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../script/">« Run as script</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.22 on <span class="colophon-date" title="Monday 25 July 2022 02:55">Monday 25 July 2022</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
